### DISCUSSION TITLE
How to dynamicly import actors /services?

### DISCUSSION THREAD

**Discussion by RafalFilipek** [statelyai/xstate#1335] [General] [answered]
Hi 👋

How to dynamically import actors /services?

Simple example: https://codesandbox.io/s/fervent-banach-pdg8q?file=/src/App.tsx

It's not working and it looks to me like it is in collision with invoking Promises.

Thanks

**Comment by Andarist** [comment, marked as answer]
It's not a collision with anything. The reason why this doesn't work is simpler: `assign` is always **synchronous** so you can't use `async` functions within it.

Notice how this is a good thing - `import()` is an async operation and it _can fail_. To provide a stellar UX/fault-tolerant logic you must handle this with appropriate error handling etc.

**Reply by davidkpiano** [reply]
Also, the idea of "async machines" isn't a first-class concept. You would have to "convert" an async machine into an invoked promise or an invoked callback. Here's an example:

```js
        actions: assign({
          a: () => {
            return spawn(async (cb, receive) => {
              const { a } = await import("./a");
              const service = interpret(a, {
                parent: ({
                  send: cb
                } as any) as Interpreter<any, any>
              }).start();

              receive(event => {
                service.send(event);
              });
            });
          }
        }),
```

https://codesandbox.io/s/inspiring-wind-h3rwb?file=/src/App.tsx:308-758
